### ISSUE TITLE
Nullable field state may be carried over unexpectedly in foreach loops

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#63478]
In the following example, we do a null test on `C.Item`, then branch backwards to a point where `c` is assigned the next element of `cArray`. It seems like the null state of `c` is "reset" here, presumably based on that assignment, but for some reason the state of nested fields is not reset.

It's possible that this behavior is intended, or perhaps preferable to the behavior where we do not warn on the first access of `c.Item`.

[SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwMQwK4pRYFAUwAIiYDiBYAKFoAEBmE+gJhIGESBvWk/ls3oBGAAwkAkgiJgSAXhIAiCIoDctPgKYsALCQCyACg4BtALokAxgEEATrYgBPAJSb+vGgK8kAZgHtbIghLAAsSQwA3CFsrElgrOwcXN28Pb3SSAHpMkgB3aJhYAHMQEgBySwA6KRkykkhHEmBSHDxKlIyraukwSoAVPwBlBFtiw2d1T06BKpqwAH5+oZGxiY6BAF91/m3fAKDQ8KiYyziYBPsnVyn0tOn+bJIYPzyC4t2vWZ6l4dGYIvGk3u/EsizmPxW/0Buy2NxIsI2QA)

```cs
#nullable enable

public class C {
    public string Item = "a";

    public void M(C[] cArray)
    {
        foreach (var c in cArray)
        {
            // warning: 'c.Item' may be null.
            c.Item.ToString();
            c.Item?.ToString();
        }
        
        foreach (var c in cArray)
        {
            // no warning
            c.Item.ToString();
            c?.Item.ToString();
        }
    }
}
```

**Comment by RikkiGibson** [comment]
Tagging @jjonescz because it's possible that this is another bug with `loopHeadState`.
